[22] VerifyError when String template expression is used as part of condition

```
interface TemplateProcessor extends java.lang.StringTemplate.Processor<Boolean, RuntimeException> {
    Boolean process(StringTemplate st);
}

public class X {
    public static void main(String argv[]) {
        TemplateProcessor STR = st -> st.interpolate().equals("abc");
        if (STR."abc") {
        	System.out.println("hello");
        }
    } 
}
```
This is supposed to print "hello, but we get this error:

```
Error: Unable to initialize main class X
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    X.main([Ljava/lang/String;)V @20: ifeq
  Reason:
    Type 'java/lang/Boolean' (current frame, stack[0]) is not assignable to integer
  Current Frame:
    bci: @20
    flags: { }
    locals: { '[Ljava/lang/String;', 'TemplateProcessor' }
    stack: { 'java/lang/Boolean' }
```

I suspect this will be a problem with other statements like 'for' and 'while' as well.

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/TemplateExpression.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/TemplateExpression.java
@@ -80,6 +80,8 @@ public class TemplateExpression extends Expression {
 		codeStream.checkcast(this.invocation.binding.returnType);
 		if (!valueRequired) {
 			codeStream.pop();
+		} else {
+			codeStream.generateImplicitConversion(this.implicitConversion);
 		}
 	}
 	@Override

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/StringTemplateTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/StringTemplateTest.java
@@ -1930,4 +1930,50 @@ s
 				"hello"
 		);
 	}
+	public void testIssue2121_02() {
+		runConformTest(
+				new String[] {
+					"X.java",
+					"""
+						interface TemplateProcessor extends java.lang.StringTemplate.Processor<Boolean, RuntimeException> {
+						    Boolean process(StringTemplate st);
+						}
+
+						public class X {
+						    public static void main(String argv[]) {
+						        TemplateProcessor STR = st -> st.interpolate().equals("abc");
+					   			int  i = 0;
+					   			while ((STR."abc") && i < 1) {
+					   				i++;
+						        	System.out.println("hello");
+						        }
+						    }
+						}
+					"""
+				},
+				"hello"
+		);
+	}
+	public void testIssue2121_03() {
+		runConformTest(
+				new String[] {
+					"X.java",
+					"""
+						interface TemplateProcessor extends java.lang.StringTemplate.Processor<Boolean, RuntimeException> {
+						    Boolean process(StringTemplate st);
+						}
+
+						public class X {
+						    public static void main(String argv[]) {
+						        TemplateProcessor STR = st -> st.interpolate().equals("abc");
+					   			for ( int i = 0; (STR."abc") && i < 1; i++) {
+						        	System.out.println("hello");
+						        }
+						    }
+						}
+					"""
+				},
+				"hello"
+		);
+	}
 }

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/StringTemplateTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/StringTemplateTest.java
@@ -1908,4 +1908,26 @@ s
 				\\{} plus \\{} equals \\{}
 				[10, 20, 30]""");
 	}
+	public void testIssue2121_01() {
+		runConformTest(
+				new String[] {
+					"X.java",
+					"""
+						interface TemplateProcessor extends java.lang.StringTemplate.Processor<Boolean, RuntimeException> {
+						    Boolean process(StringTemplate st);
+						}
+
+						public class X {
+						    public static void main(String argv[]) {
+						        TemplateProcessor STR = st -> st.interpolate().equals("abc");
+						        if (STR."abc") {
+						        	System.out.println("hello");
+						        }
+						    }
+						}
+					"""
+				},
+				"hello"
+		);
+	}
 }

